fix DPO data collator

nrailg opened the pull request:
Fix problems described in [this issue 907](https://github.com/huggingface/trl/issues/907).

Testing code: https://github.com/nrailg/trl-dpo-alpaca-farm-demo

@kashif @lvwerra

younesbelkada commented on trl/trainer/utils.py:
```diff
@@ -318,31 +318,32 @@ def tokenize_batch_element(
-                0 if i in eos_indices_rejected else p for i, p in enumerate(rejected_tokens["attention_mask"])
-            ]
-            rejected_tokens["attention_mask"] = new_attention_mask_r
+            assert isinstance(prompt, str)
```
```suggestion
            if not isinstance(prompt, str):
                raise ValueError(f"prompt should be an str, got {type(prompt)}")
```
Can you do something similar for all `assert` below? 🙏

younesbelkada reviewed (commented):
Looks great, thanks! I left one suggestion that can be propagated across some lines below, what do you think? Also let's wait to hear from @kashif to see if this fix is the right fix

nrailg commented:
> Looks great, thanks! I left one suggestion that can be propagated across some lines below, what do you think? Also let's wait to hear from @kashif to see if this fix is the right fix

Upgraded.

nrailg commented:
@kashif Hello, I'll be regretful if you could spare some time and review this PR.

kashif commented:
@nrailg thanks! having a look now... so would a user need to create the DPO type dataset differently due to this?

nrailg commented:
> @nrailg thanks! having a look now... so would a user need to create the DPO type dataset differently due to this?

NO.

kashif commented:
@nrailg  yes this looks great and brings the tokenization more in line with what is done so thank you! I am currently refactoring the DPO collator so i would potentially need to add these changes to that, if you want, I can add you as a collaborator if you want to add it there since I have made helper function for the tokenization, for which we can also add tests? what do you think?

kashif commented on trl/trainer/utils.py:
```diff
@@ -318,31 +318,35 @@ def tokenize_batch_element(
+            prompt_tokens = self.tokenizer(prompt, add_special_tokens=False)
+            prompt_input_ids = prompt_tokens["input_ids"]
+            prompt_attention_mask = prompt_tokens["attention_mask"]
+            assert len(prompt_input_ids) == len(prompt_attention_mask)
```
```suggestion
            if len(prompt_input_ids) != len(prompt_attention_mask):
                raise ValueError("Prompt input ids and attention mask should have the same length.")
```

kashif commented on trl/trainer/utils.py:
```diff
@@ -318,31 +318,35 @@ def tokenize_batch_element(
+            if not isinstance(chosen, str):
+                raise ValueError(f"chosen should be an str but got {type(chosen)}")
+            chosen_tokens = self.tokenizer(prompt + chosen, add_special_tokens=False)
+            assert prompt_input_ids == chosen_tokens["input_ids"][: len(prompt_input_ids)]
```
```suggestion
            if prompt_input_ids != chosen_tokens["input_ids"][: len(prompt_input_ids)]:
                raise ValueError("Prompt input ids and chosen input ids should be the same up to the prompt.")
```

kashif commented on trl/trainer/utils.py:
```diff
@@ -318,31 +318,35 @@ def tokenize_batch_element(
+            if not isinstance(rejected, str):
+                raise ValueError(f"rejected should be an str but got {type(rejected)}")
+            rejected_tokens = self.tokenizer(prompt + rejected, add_special_tokens=False)
+            assert prompt_input_ids == rejected_tokens["input_ids"][: len(prompt_input_ids)]
```
```suggestion
            if prompt_input_ids != rejected_tokens["input_ids"][: len(prompt_input_ids)]:
                raise ValueError("Prompt input ids and rejected input ids should be the same up to the prompt.")
```

kashif commented:
@nrailg  are your tests passing?

nrailg commented:
> @nrailg are your tests passing? i think for the gpt2 tokenizer i am using the tests will not pass for you

1. I ran the stackllama2 demo, and finished training my own demo.
2. I didn't find any test case related to DPO data collator by grep "DPODataCollator", so I didn't run `make test`.

nrailg commented:
> @nrailg yes this looks great and brings the tokenization more in line with what is done so thank you! I am currently refactoring the DPO collator so i would potentially need to add these changes to that, if you want, I can add you as a collaborator if you want to add it there since I have made helper function for the tokenization, for which we can also add tests? what do you think? see #885

Glad to help. Give me sometime to figuring out what you changed.

kashif commented:
I am still testing on my side, @nrailg, and I am getting an issue where in my eval dataset I get an error in the embedding layer... still trying to figure out why... can you kindly test on your side? we fixed some edge cases not covered by your original PR here too

nrailg commented:
> I am still testing on my side, @nrailg, and I am getting an issue where in my eval dataset I get an error in the embedding layer... still trying to figure out why... can you kindly test on your side? we fixed some edge cases not covered by your original PR here too

we fixed some edge cases not covered by your original PR here too
-------------------------------------------------------------------
Thank you so much. I'll try figuring out how to run `make test` in my company's weird gpu cluster.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/trl/pr_932). All of your documentation changes will be reflected on that endpoint.
